imap_processing/ccsds/excel_to_xtce.py

```diff
@@ -0,0 +1,426 @@
+        """
+        parameter = Et.SubElement(self._parameter_set, "xtce:Parameter")
+        # Combine the packet name and mnemonic to create a unique parameter name
+        name = f"{row['packetName']}_{row['mnemonic']}"
+        parameter.attrib["name"] = name
+        # UINT8, ...
+        parameter.attrib["parameterTypeRef"] = name
+
+        # Add descriptions if they exist
+        if pd.notna(row.get("shortDescription")):
+            parameter.attrib["shortDescription"] = row.get("shortDescription")
+        if pd.notna(row.get("longDescription")):
+            description = Et.SubElement(parameter, "xtce:LongDescription")
+            description.text = row.get("longDescription")
+
+        # Add the parameterTypeRef for this row
+        if "UINT" in row["dataType"]:
+            parameter_type = Et.SubElement(
+                self._parameter_type_set, "xtce:IntegerParameterType"
+            )
+            parameter_type.attrib["name"] = name
+            parameter_type.attrib["signed"] = "false"
+
+            encoding = Et.SubElement(parameter_type, "xtce:IntegerDataEncoding")
+            encoding.attrib["sizeInBits"] = str(row["lengthInBits"])
+            encoding.attrib["encoding"] = "unsigned"
+
+        elif any(x in row["dataType"] for x in ["SINT", "INT"]):
+            parameter_type = Et.SubElement(
+                self._parameter_type_set, "xtce:IntegerParameterType"
+            )
+            parameter_type.attrib["name"] = name
+            parameter_type.attrib["signed"] = "true"
+            encoding = Et.SubElement(parameter_type, "xtce:IntegerDataEncoding")
+            encoding.attrib["sizeInBits"] = str(row["lengthInBits"])
+            encoding.attrib["encoding"] = "signed"
+
+        elif "BYTE" in row["dataType"]:
+            parameter_type = Et.SubElement(
+                self._parameter_type_set, "xtce:BinaryParameterType"
+            )
+            parameter_type.attrib["name"] = name
+
+            encoding = Et.SubElement(parameter_type, "xtce:BinaryDataEncoding")
+            encoding.attrib["bitOrder"] = "mostSignificantBitFirst"
+
+            size_in_bits = Et.SubElement(encoding, "xtce:SizeInBits")
+            if final_two_rows:
+                # If it is a byte field in the final two rows, consider it a
+                # dynamic value. (there is sometimes an extra checksum in the final row)
+                dynamic_value = Et.SubElement(size_in_bits, "xtce:DynamicValue")
+                param_ref = Et.SubElement(dynamic_value, "xtce:ParameterInstanceRef")
+                param_ref.attrib["parameterRef"] = "PKT_LEN"
+                linear_adjustment = Et.SubElement(
+                    dynamic_value, "xtce:LinearAdjustment"
+                )
+                linear_adjustment.attrib["slope"] = str(8)
+                # Intercept is the start bit of this parameter - (header bytes - 1) * 8
+                linear_adjustment.attrib["intercept"] = str(-(row["startBit"] - 40))
```

Comment:
I like having this be automatic, but the assumption about the last 2 rows being dynamic seems like it's not right? I guess if we assume the last row is always dynamic and it isn't then it works the same way, so that part seems fine. But why the last two, if we can only have one dynamic length variable?

Reply:
> But why the last two, if we can only have one dynamic length variable?

Is this true? If so, I think we could just make all BYTE entries be dynamic and it will be easier than restricting it to the final two entries.

Testing with Hi and SWAPI, this did appear to work with both the final row and the second-to-last row being correct.

Reply:
It was my understanding that there could only be one variable-length field in the XTCE. All my knowledge of variable length fields basically comes from [this](https://lasp.slack.com/archives/C05CRQL2LJF/p1695745620068189) slack thread though. [Here](https://lasp.slack.com/archives/C05CRQL2LJF/p1694535292813049) is also some info.